src/bsoncxx/json.cpp

```diff
@@ -43,12 +43,28 @@ void bson_free_deleter(std::uint8_t* ptr) {
 
 }  // namespace
 
-std::string BSONCXX_CALL to_json(document::view view) {
+enum class jsonMode { LEGACY, CANONICAL, RELAXED };
+
+std::string _to_json_helper(document::view view, jsonMode mode)
+{
     bson_t bson;
     bson_init_static(&bson, view.data(), view.length());
 
     size_t size;
-    auto result = bson_as_json(&bson, &size);
+    char  *result;
+
+    switch (mode) {
```

Comment:
Since all these variations call functions with the same signature, why don't you just pass the function you want to call in here?

```
std::string to_json_helper(document::view view, decltype(&bson_as_json) converter) {
   bson_t bson;
   bson_init_static(...);
   size_t size;
   auto result = converter(&bson, &size);
   if (!result)
       throw...
   return {result, size};
}
```

Then `to_canonical_json` is just:

```
... {
    return to_json_helper(view, &bson_as_canonical_extended_json);
}
```

And you don't need the enum at all.

Reply:
Likewise, if we opt for a single method that takes an enum argument we could reduce the code dupe.